CI: WaitForDeploy/Daemonset

raybejjani opened the pull request:
WaitForDeployment uses the number of pods configured in k8s to determine
if the deploy is ready. This should be resilient to still-terminated
pods left over from other tests and better reflect what the k8s
scheduler thinks is true.

WaitForDaemonSetReady uses the k8s expected pods to determine that pods
are ready. This sidesteps passing a count to WaitforNPods or passing 0
to cause it to determine the number of pods from those currently
running.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8110)
<!-- Reviewable:end -->

raybejjani committed:
CI: Add/Use WaitforDeploy & ExpectDeployReady

WaitForDeployment uses the number of pods configured in k8s to determine
if the deploy is ready. This should be resilient to still-terminated
pods left over from other tests and better reflect what the k8s
scheduler thinks is true.

ExpectCiliumReady, ExpectCiliumOperatorReady and ExpectETCDOperatorReady
have been switched over to use the new scheme. Other WaitforNPods
invocations can also be switched in the future.

raybejjani committed:
CI: Add WaitForDaemonSetReady & ExpectDaemonSetReady

WaitForDaemonSetReady uses the k8s expected pods to determine that pods
are ready. This sidesteps passing a count to WaitforNPods or passing 0
to cause it to determine the number of pods from those currently
running.

Utility functions changed:
- ExpectCiliumPreFlightInstallReady
- ExpectCiliumOperatorReady
- ExpectCiliumRunning
- ExpectETCDOperatorReady

Tests changed:
- K8sChaosTest
- NightlyEpsMeasurement
- K8sServicesTest

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on test/k8sT/Policies.go:
```diff
@@ -869,13 +871,12 @@ EOF`, k, v)
-			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", helpers.HelperTimeout)
-			Expect(err).To(BeNil(),
-				"testapp pods are not ready after timeout in %q namespace", helpers.DefaultNamespace)
+			ExpectDeployReady(kubectl, secondNS, "app1", helpers.HelperTimeout)
```
Can you change these and other string literals to use `helpers.AppN` where "N" is one of {1, 2, 3}?

ianvernon commented on test/k8sT/assertionHelpers.go:
```diff
@@ -38,18 +38,15 @@ func ExpectKubeDNSReady(vm *helpers.Kubectl) {
 func ExpectCiliumOperatorReady(vm *helpers.Kubectl) {
-	err := vm.WaitforPods(helpers.KubeSystemNamespace, "-l name=cilium-operator", longTimeout)
-	ExpectWithOffset(1, err).Should(BeNil(), "Cilium operator was not able to get into ready state")
+	ExpectDeployReady(vm, "kube-system", "cilium-operator", longTimeout)
```
This PR has a lot of changes - which may make it hard to debug if it fails - maybe tease out stuff into separate commits and add the base infrastructure along with adding something like `ExpectCIliumOperatorReady` as the initial consumer of `ExpectDeployReady` since it's used in all K8s tests? This isn't strictly necessary, but may make your life easier for getting this merged since this commit contains a lot changes.

ianvernon commented on test/k8sT/assertionHelpers.go:
```diff
@@ -104,6 +91,16 @@ func ExpectCiliumPreFlightInstallReady(vm *helpers.Kubectl) {
 	Expect(err).To(BeNil(), "cilium pre-flight check is not ready after timeout, pods status:\n %s", warningMessage)
 }
 
+func ExpectDeployReady(vm *helpers.Kubectl, namespace, name string, timeout time.Duration) {
```
+1

ianvernon commented on test/k8sT/Chaos.go:
```diff
@@ -55,10 +55,8 @@ var _ = Describe("K8sChaosTest", func() {
-			err := kubectl.WaitforPods(
-				helpers.DefaultNamespace, fmt.Sprintf("-l zgroup=testDS"), helpers.HelperTimeout)
-			Expect(err).Should(BeNil(), "Pods are not ready after timeout")
+			ExpectDaemonSetReady(kubectl, "default", "testds", helpers.HelperTimeout)
```
use `helpers.DefaultNamespace`

raybejjani replied:
I even found an older instance where it can be changed too!

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -596,6 +596,28 @@ func (kub *Kubectl) WaitforDeployReady(namespace, name string, timeout time.Dura
+
+	body := func() bool {
+		ds := apps_v1.DaemonSet{}
+		err := kub.Exec(fmt.Sprintf("%s -n %s get daemonset %s -o json", KubectlCmd, namespace, name)).Unmarshal(&ds)
```
The `*CmdRes` returned by `kub.Exec` may be nil, which would result in a panic.

raybejjani replied:
I'll check for nil explicitly but where do you see it being nil? It seems to be unconditionally created and returned in ExecContext: https://github.com/cilium/cilium/blob/master//test/helpers/node.go#L202-L208.

ianvernon replied:
Ah, okay, if we are guaranteed we will return a non-nil pointer, then this is fine!

ianvernon commented on test/k8sT/Services.go:
```diff
@@ -120,6 +120,8 @@ var _ = Describe("K8sServicesTest", func() {
 		BeforeEach(func() {
 			res := kubectl.Apply(demoYAML)
 			res.ExpectSuccess("unable to apply %s", demoYAML)
+			ExpectDaemonSetReady(kubectl, "default", "testds", helpers.HelperTimeout)
```
Why are these additions needed here?

raybejjani replied:
Nothing direct. My reasoning was that `res.ExpectSuccess("unable to apply %s", demoYAML)` implied that the pods were added but nothing checked that they were actually up in the tests. I'll see if this can go into a separate commit.

ianvernon commented on test/k8sT/Services.go:
```diff
@@ -160,6 +162,8 @@ var _ = Describe("K8sServicesTest", func() {
 		BeforeAll(func() {
 			res := kubectl.Apply(demoYAML)
```
Perhaps refactor this and the above on line 121-124 into a separate function to reduce code duplication.

raybejjani replied:
I started down this road but I ended up with more code than before. It's trivial to factor out the code itself but then I noticed that there is a cleanup section, but it is distinct between the two contexts. It felt a little odd to wrap the creation but leave the deletion open. (There was also the oddness of the BeforeAll taking the wrapper directly, and that looked weird compared to all the other things). Since these were different contexts (and different AfterAll vs AfterEach) I left it as it is.

ianvernon replied:
Makes sense, thanks for explaining!

ianvernon commented on test/k8sT/assertionHelpers.go:
```diff
@@ -104,6 +91,16 @@ func ExpectCiliumPreFlightInstallReady(vm *helpers.Kubectl) {
+	Expect(err).To(BeNil(), "Deploy %s/%s not ready after timeout:\n %s", namespace, name, err)
+}
+
+func ExpectDaemonSetReady(vm *helpers.Kubectl, namespace, name string, timeout time.Duration) {
```
+1

ianvernon reviewed (changes requested):
Looks good overall, some minor nits. I think it will be quite hard to get this plumbed in everywhere in one PR - perhaps a more iterative approach is best? Phasing in the function and its usage in one test, then slowly seeing what needs to be fixed. This is up to you, since the changes look fine, but it may be harder to debug / get this merged if the tests are failing :/

raybejjani committed:
CI: K8sServicesTest consistenly uses global DefaultNamespace

This test mixed literal "default" and the global variable. It now uses
only the global variable.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on test/k8sT/Services.go:
```diff
@@ -120,6 +120,8 @@ var _ = Describe("K8sServicesTest", func() {
 		BeforeEach(func() {
 			res := kubectl.Apply(demoYAML)
 			res.ExpectSuccess("unable to apply %s", demoYAML)
+			ExpectDaemonSetReady(kubectl, helpers.DefaultNamespace, "testds", helpers.HelperTimeout)
```
There is no `DaemonSet` called `testds` in this test - perhaps you  meant `ExpectDeployReady`, as `demo.yaml` contains deployments?

ianvernon commented on test/k8sT/Chaos.go:
```diff
@@ -55,10 +55,8 @@ var _ = Describe("K8sChaosTest", func() {
-			err := kubectl.WaitforPods(
-				helpers.DefaultNamespace, fmt.Sprintf("-l zgroup=testDS"), helpers.HelperTimeout)
-			Expect(err).Should(BeNil(), "Pods are not ready after timeout")
+			ExpectDaemonSetReady(kubectl, helpers.DefaultNamespace, "testds", helpers.HelperTimeout)
```
Perhaps factor out `testds` and `testclient` into constants?

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -295,13 +295,15 @@ var _ = Describe("NightlyEpsMeasurement", func() {
 			kubectl.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
 			manifest := helpers.ManifestGet(netcatDsManifest)
 			kubectl.Apply(manifest).ExpectSuccess("Cannot apply netcat ds")
+			ExpectDaemonSetReady(kubectl, helpers.DefaultNamespace, "netcatds", helpers.HelperTimeout)
```
Factor `netcatds` into a constant.

ianvernon commented on Gopkg.lock:
```diff
@@ -1887,6 +1887,7 @@
     "gopkg.in/fsnotify.v1",
     "gopkg.in/natefinch/lumberjack.v2",
     "gopkg.in/yaml.v2",
+    "k8s.io/api/apps/v1",
```
Nit: should this be in the prior commit vs. this one?

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -574,6 +574,37 @@ func (kub *Kubectl) NamespaceDelete(name string) *CmdRes {
+		specReplicas = deploy.Status.Replicas
+		currentReplicas = deploy.Status.ReadyReplicas
+
+		return currentReplicas >= specReplicas
```
Why not use `==` instead of `>`?
